python/src/main/python/yaml-template/main.py

```diff
@@ -13,18 +13,52 @@
 # License for the specific language governing permissions and limitations under
 # the License.
 #
-
+#
 import argparse
+import json
 import logging
+import os
+from argparse import Namespace
+
+import jinja2
 
+from apache_beam.io.filesystems import FileSystems
 from apache_beam.yaml import cache_provider_artifacts
 from apache_beam.yaml import main
 
 
-def run(argv=None):
+def _configure_parser(argv):
   parser = argparse.ArgumentParser()
-  _, pipeline_args = parser.parse_known_args(argv)
-  pipeline_args += ['--sdk_location=container']
+  parser.add_argument(
+    '--jinja_variables',
+    default=None,
+    type=json.loads,
+    help='A json dict of variables used when invoking the jinja preprocessor '
+         'on the provided yaml pipeline.')
+  return parser.parse_known_args(argv)
+
+
+class _BeamFileIOLoader(jinja2.BaseLoader):
+  def get_source(self, environment, path):
+    source = FileSystems.open(path).read().decode()
+    return source, path, lambda: True
+
+
+def run(argv=None):
+  known_args, pipeline_args = _configure_parser(argv)
+  known_args = Namespace(
+    **vars(main._configure_parser(argv)[0]), **vars(known_args))
+  pipeline_yaml = (
+    jinja2.Environment(
+      undefined=jinja2.StrictUndefined, loader=_BeamFileIOLoader())
+    .from_string(main._pipeline_spec_from_args(known_args))
+    .render(**known_args.jinja_variables or {}))
+  pipeline_yaml = os.linesep.join(
+    [s for s in pipeline_yaml.splitlines() if s.strip()])
+
+  pipeline_args += [
+    '--sdk_location=container', f'--yaml_pipeline={pipeline_yaml}'
```

Comment:
Why are we overrideing sdk_location?

Reply:
Recommended in https://cloud.google.com/dataflow/docs/guides/build-container-image#preinstall_using_a_dockerfile since sdk is already pre-installed in image. Supposedly speeds up startup, but not sure how significantly